daemon/info.go

```diff
@@ -62,6 +62,7 @@ func (daemon *Daemon) SystemInfo() *system.Info {
 		NoProxy:            getConfigOrEnv(cfg.NoProxy, "NO_PROXY", "no_proxy"),
 		LiveRestoreEnabled: cfg.LiveRestoreEnabled,
 		Isolation:          daemon.defaultIsolation,
+		CDISpecDirs:        cfg.CDISpecDirs,
```

Comment:
@thaJeztah this is always equal to the value in the config file and is never overridden by the CLI values. (Hence the failing tests that were added).

Note that this  matches what is done for `experimental` which can also be set via the command line. Is the expectation that the command line options be shown in the `docker info` output, or shoudl the config file value always be shown?

If it is the former, do you have any pointers on where to get started? I would assume that we need to add members to the `Daemon` type itself similar to what is done for `genericResources` and update these when parsing the config options when the daemon is started.

Reply:
Thanks for the pointers @corhere. You're right that I was not correctly interpreting the failed test output.

From my side I would expect `CDISpecDir` settings specified in the CLI or in the daemon.json should override the defaults and not be appended. I have updated `loadDaemonCliConfig` to check for a `nil` config option and set the defaults if this is not set.

Do you have any thoughts / views on whether an empty (non-nil) slice should also have the defaults applied?

Reply:
I think it would be more surprising for a daemon.json `{"cdi-spec-dirs": []}` to result in the defaults being applied. I would intuitively expect it to override the defaults, such that the configured CDI spec dirs would be empty. Perhaps some operator/administrator would want to disable CDI support on their engine for whatever reason; this would give them a more direct and explicit way to do so than pointing the engine at a placeholder path. (Read: I agree with your choice as currently implemented.) I would like to see that case covered by a test to assure that we don't regress on that edge case.

On the CLI arguments side, perhaps `dockerd --cdi-spec-dir=` (i.e. empty option value) could be used to achieve the same? I'm not sure how easy that would be to pull off with pflag and our config framework, however.

Reply:
Thanks for the feedback. I was also thinking that setting `{"cdi-spec-dirs": []}` should disable CDI without requiring an explicit feature flag. I like having this overridable as a command line option too -- by specifying a single empty option as you showed. I will add some logic to do this and cover the various cases in additional unit tests.

Before I dig through additional existing tests, do you have pointers handy on tests that set `config.json` values explicitly?

Reply:
I couldn't find any integration test helpers for starting a daemon under test with a config file. Looks like all the tests which need it open-code writing the config to a file and calling `d.Start(t, "--config-file", path)`. https://github.com/moby/moby/blob/afd4805278b488c2dbf53394660469d47db79398/integration/system/ping_test.go#L124-L135

You could write the config file to `d.RootDir()` as the test above does, or use `t.TempDir()`.